File: Assets/Bundles/Shared/Resources/TS/Util/SpringTween.lua
-- Compiled with unity-ts v2.1.0-75
local Spring = require("Shared/TS/Util/Spring").Spring
local EPSILON = 0.0001
local SpringTween
do
	SpringTween = setmetatable({}, {
		__tostring = function()
			return "SpringTween"
		end,
	})
	SpringTween.__index = SpringTween
	function SpringTween.new(...)
		local self = setmetatable({}, SpringTween)
		return self:constructor(...) or self
	end
	function SpringTween:constructor(position, angularFrequency, minDuration)
		self.minDuration = minDuration
		self.elapsedTime = 0
		self.spring = Spring.new(position, angularFrequency)
	end
	function SpringTween:setGoal(goal)
		self.elapsedTime = 0
		self.spring.goal = goal
	end
	function SpringTween:impulse(impulse)
		self.elapsedTime = 0
		self.spring:impulse(impulse)
	end
	function SpringTween:resetTo(position)
		self.elapsedTime = 0
		self.spring:resetTo(position)
	end
	function SpringTween:update(deltaTime)
		self.elapsedTime += deltaTime
		local springPos = self.spring:update(deltaTime)
		return springPos, self.elapsedTime >= self.minDuration and self.spring.velocity.magnitude < EPSILON
	end
end
return {
	SpringTween = SpringTween,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/Util/SpringTween.ts
import { Spring } from "./Spring";

const EPSILON = 0.0001;

export class SpringTween {
	private readonly spring: Spring;

	private elapsedTime = 0;

	constructor(position: Vector3, angularFrequency: number, private readonly minDuration: number) {
		this.spring = new Spring(position, angularFrequency);
	}

	public SetGoal(goal: Vector3) {
		this.elapsedTime = 0;
		this.spring.goal = goal;
	}

	public Impulse(impulse: Vector3) {
		this.elapsedTime = 0;
		this.spring.Impulse(impulse);
	}

	public ResetTo(position: Vector3) {
		this.elapsedTime = 0;
		this.spring.ResetTo(position);
	}

	public Update(deltaTime: number): LuaTuple<[position: Vector3, isDone: boolean]> {
		this.elapsedTime += deltaTime;
		const springPos = this.spring.Update(deltaTime);
		return $tuple(springPos, this.elapsedTime >= this.minDuration && this.spring.velocity.magnitude < EPSILON);
	}
}
